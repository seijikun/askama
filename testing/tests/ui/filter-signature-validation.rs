mod missing_required_args {
    #[askama::filter_fn]
    pub fn filter0() -> askama::Result<String> {}

    #[askama::filter_fn]
    pub fn filter1(input: usize) -> askama::Result<String> {}

    #[askama::filter_fn]
    pub fn filter2(_: &dyn askama::Values) -> askama::Result<String> {}
}

mod lifetime_args {
    #[askama::filter_fn]
    pub fn filter0<'a>(input: usize, _: &dyn askama::Values, arg: &'a ()) -> askama::Result<String> {}
}

mod const_generic_args {
    #[askama::filter_fn]
    pub fn filter0<const T: bool>(input: usize, _: &dyn askama::Values) -> askama::Result<String> {}
}

mod generic_default {
    #[askama::filter_fn]
    pub fn filter0<T = f64>(input: usize, _: &dyn askama::Values, arg: T) -> askama::Result<String> {}
}

mod missing_rettype {
    #[askama::filter_fn]
    pub fn filter0(input: usize, _: &dyn askama::Values) {}
}

mod argument_destructuring {
    pub struct Wrapper(i64);
    #[askama::filter_fn]
    pub fn filter0(input: usize, _: &dyn askama::Values, Wrapper(arg): Wrapper) -> askama::Result<String> {}
}

mod argument_impl_generic {
    #[askama::filter_fn]
    pub fn filter0(input: usize, _: &dyn askama::Values, arg: impl std::fmt::Display) -> askama::Result<String> {}
}

mod optional_args_before_required {
    #[askama::filter_fn]
    pub fn filter0(input: usize, _: &dyn askama::Values, #[optional(1337)] opt_arg: usize, req_arg: usize) -> askama::Result<String> {}
}

mod optional_arg_generics {
    #[askama::filter_fn]
    pub fn filter0(input: usize, _: &dyn askama::Values, #[optional(1337)] opt_arg: impl std::fmt::Display) -> askama::Result<String> {}

    #[askama::filter_fn]
    pub fn filter1<T: Copy>(input: usize, _: &dyn askama::Values, #[optional(1337usize)] opt_arg: T) -> askama::Result<String> {}

    #[askama::filter_fn]
    pub fn filter2<T: Copy>(input: usize, _: &dyn askama::Values, #[optional(1337usize)] opt_arg: Option<T>) -> askama::Result<String> {}
}

pub fn main() {}