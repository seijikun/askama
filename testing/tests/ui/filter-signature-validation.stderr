error: Filter function missing required input and environment arguments. Example: `fn filter0(_: &dyn std::fmt::Display, _: &dyn askama::Values) -> askama::Result<String>`
 --> tests/ui/filter-signature-validation.rs:3:19
  |
3 |     pub fn filter0() -> askama::Result<String> {}
  |                   ^

error: Filter function missing required environment argument. Example: `fn filter0(_: &dyn std::fmt::Display, _: &dyn askama::Values) -> askama::Result<String>`
 --> tests/ui/filter-signature-validation.rs:6:19
  |
6 |     pub fn filter1(input: usize) -> askama::Result<String> {}
  |                   ^

error: Filter function missing required environment argument. Example: `fn filter0(_: &dyn std::fmt::Display, _: &dyn askama::Values) -> askama::Result<String>`
 --> tests/ui/filter-signature-validation.rs:9:19
  |
9 |     pub fn filter2(_: &dyn askama::Values) -> askama::Result<String> {}
  |                   ^

error: Lifetime generics are currently not supported for filters
  --> tests/ui/filter-signature-validation.rs:14:20
   |
14 |     pub fn filter0<'a>(input: usize, _: &dyn askama::Values, arg: &'a ()) -> askama::Result<String> {}
   |                    ^^

error: Const generics are currently not supported for filters
  --> tests/ui/filter-signature-validation.rs:19:20
   |
19 |     pub fn filter0<const T: bool>(input: usize, _: &dyn askama::Values) -> askama::Result<String> {}
   |                    ^^^^^

error: Filter functions don't support generic parameter defaults
  --> tests/ui/filter-signature-validation.rs:24:24
   |
24 |     pub fn filter0<T = f64>(input: usize, _: &dyn askama::Values, arg: T) -> askama::Result<String> {}
   |                        ^^^

error: Filter function is missing return type
  --> tests/ui/filter-signature-validation.rs:29:56
   |
29 |     pub fn filter0(input: usize, _: &dyn askama::Values) {}
   |                                                        ^

error: Only conventional function arguments are supported
  --> tests/ui/filter-signature-validation.rs:35:58
   |
35 |     pub fn filter0(input: usize, _: &dyn askama::Values, Wrapper(arg): Wrapper) -> askama::Result<String> {}
   |                                                          ^^^^^^^

error: Impl generics are currently not supported for filters
  --> tests/ui/filter-signature-validation.rs:40:63
   |
40 |     pub fn filter0(input: usize, _: &dyn askama::Values, arg: impl std::fmt::Display) -> askama::Result<String> {}
   |                                                               ^^^^

error: All required arguments must appear before any optional ones
  --> tests/ui/filter-signature-validation.rs:45:92
   |
45 |     pub fn filter0(input: usize, _: &dyn askama::Values, #[optional(1337)] opt_arg: usize, req_arg: usize) -> askama::Result<String> {}
   |                                                                                            ^^^^^^^

error: Impl generics are currently not supported for filters
  --> tests/ui/filter-signature-validation.rs:50:85
   |
50 |     pub fn filter0(input: usize, _: &dyn askama::Values, #[optional(1337)] opt_arg: impl std::fmt::Display) -> askama::Result<String> {}
   |                                                                                     ^^^^

error: Optional arguments must not use generic parameters
  --> tests/ui/filter-signature-validation.rs:53:99
   |
53 |     pub fn filter1<T: Copy>(input: usize, _: &dyn askama::Values, #[optional(1337usize)] opt_arg: T) -> askama::Result<String> {}
   |                                                                                                   ^

error: Optional arguments must not use generic parameters
  --> tests/ui/filter-signature-validation.rs:56:106
   |
56 |     pub fn filter2<T: Copy>(input: usize, _: &dyn askama::Values, #[optional(1337usize)] opt_arg: Option<T>) -> askama::Result<String> {}
   |                                                                                                          ^
